boundary/analyze.py

```diff
@@ -328,12 +329,26 @@ class dotdict(dict):
         'callback': set(),
         'sdcr_fns': set(),
         'interface': set(),
+        'weak': set(),
     })
 
     edges = []
     decls = {}
     hdr_sym = {'fn': list(), 'var': list()}
 
+    for meta in metas:
+        for fn in meta['fn']:
+            fn = dotdict(fn)
+            if fn.weak:
+                """arch_kexec_apply_relocations_add was defined as weak
+                symbol 2 times in the kernel, and this bug is fixed in
+                "kexec_file: Allow archs to handle special regions while
+                locating memory hole". We can only workadround it."""
+                if fn.name != 'arch_kexec_apply_relocations_add':
+                    assert fn.name not in global_fn_dict, \
+                        "Found multiple weak symbol %s" % fn.name
+                global_fn_dict[fn.name] = fn.file
+
```

Comment:
Can we put this logic to the next "for meta in metas" loop ?  Adding extra pass looks bloated to me.

Reply:
@chenshanpei I agree. And according to testing report from @dtcccc , this PR is very unstable. So it needs to be re-designed.